src/compiler/watchPublic.ts

```diff
@@ -609,9 +609,13 @@ namespace ts {
             }
         }
 
-        function getSourceVersion(path: Path): string | undefined {
+        function getSourceVersion(path: Path, readFileWithCache: (fileName: string) => string | undefined): string | undefined {
             const hostSourceFile = sourceFilesCache.get(path);
-            return !hostSourceFile || !hostSourceFile.version ? undefined : hostSourceFile.version;
+            if (!hostSourceFile) return undefined;
+            if (hostSourceFile.version) return hostSourceFile.version;
+            // Read file and get new version
+            const text = readFileWithCache(path);
+            return text ? (compilerHost.createHash || generateDjb2Hash)(text) : undefined;

```

Comment:
There's an edge case here, right? If the file text is empty, then you still have a valid version.